core/src/main/java/org/apache/iceberg/UpdateRequirement.java

```diff
@@ -0,0 +1,371 @@
+      updates.add(update);
+
+      // add requirements based on the change
+      if (update instanceof MetadataUpdate.SetSnapshotRef) {
+        update((MetadataUpdate.SetSnapshotRef) update);
+      } else if (update instanceof MetadataUpdate.AddSchema) {
+        update((MetadataUpdate.AddSchema) update);
+      } else if (update instanceof MetadataUpdate.SetCurrentSchema) {
+        update((MetadataUpdate.SetCurrentSchema) update);
+      } else if (update instanceof MetadataUpdate.AddPartitionSpec) {
+        update((MetadataUpdate.AddPartitionSpec) update);
+      } else if (update instanceof MetadataUpdate.SetDefaultPartitionSpec) {
+        update((MetadataUpdate.SetDefaultPartitionSpec) update);
+      } else if (update instanceof MetadataUpdate.SetDefaultSortOrder) {
+        update((MetadataUpdate.SetDefaultSortOrder) update);
+      }
+
+      return this;
+    }
+
+    private void update(MetadataUpdate.SetSnapshotRef setRef) {
+      // require that the ref is unchanged from the base
+      String name = setRef.name();
+      // add returns true the first time the ref name is added
+      boolean added = changedRefs.add(name);
+      if (added && base != null && !isReplace) {
+        SnapshotRef baseRef = base.ref(name);
+        // require that the ref does not exist (null) or is the same as the base snapshot
+        requireRefSnapshotId(name, baseRef != null ? baseRef.snapshotId() : null);
+      }
+    }
+
+    private void update(MetadataUpdate.AddSchema update) {
+      if (!addedSchema) {
+        if (base != null) {
+          requireLastAssignedFieldId(base.lastColumnId());
+        }
+        this.addedSchema = true;
+      }
+    }
+
+    private void update(MetadataUpdate.SetCurrentSchema update) {
+      if (!setSchemaId) {
+        if (base != null && !isReplace) {
+          // require that the current schema has not changed
+          requireCurrentSchemaId(base.currentSchemaId());
+        }
+        this.setSchemaId = true;
+      }
+    }
+
+    private void update(MetadataUpdate.AddPartitionSpec update) {
+      if (!addedSpec) {
+        if (base != null) {
+          requireLastAssignedPartitionId(base.lastAssignedPartitionId());
+        }
+        this.addedSpec = true;
+      }
+    }
+
+    private void update(MetadataUpdate.SetDefaultPartitionSpec update) {
+      if (!setSpecId) {
+        if (base != null && !isReplace) {
+          // require that the default spec has not changed
+          requireDefaultSpecId(base.defaultSpecId());
+        }
+        this.setSpecId = true;
+      }
+    }
+
+    private void update(MetadataUpdate.SetDefaultSortOrder update) {
+      if (!setOrderId) {
+        if (base != null && !isReplace) {
+          // require that the default write order has not changed
+          requireDefaultSortOrderId(base.defaultSortOrderId());
+        }
+        this.setOrderId = true;
+      }
+    }
+
+    public List<UpdateRequirement> build() {
```

Comment:
I don't think the builder makes much sense when these classes are public. The builder was originally creating the `UpdateTableRequest`, which needed a list of both changes and requirements. Now, this is only producing a list of requirements from a list of changes so the builder pattern is unnecessary.

I think the cleanest path forward is to try to match the `MetadataUpdate` API here. Let's keep the constructors for these classes public like `MetadataUpdate` classes and add logic for converting a collection of updates into a collection of requirements elsewhere.

Reply:
I've made all the constructors public in `UpdateRequirement` and added `UpdateRequirements` with static methods that would then produce a `List<UpdateRequirement` from a `List<MetadataUpdate>`. 
We don't expose a builder anymore in the API, but `UpdateRequirements` still uses one internally, since it needs to maintain some state when producing requirements from a given list of updates.